core/primitives/src/cpu.rs

```diff
@@ -0,0 +1,97 @@
+/// Try to ensure that the required CPU features are available and enabled for the current CPU.
+///
+/// Note that this is a best-effort kind of thing. If these features are already enabled when
+/// compiling the program, then we may very well end up executing some unsupported instruction
+/// before we even get to executing this code.
+pub fn ensure_cpu_compatibility(force: bool) {
+    if cfg!(near_production) || force {
+        imp::ensure_cpu_compatibility();
+    } else {
+        // Intentionally empty: allow execution without verifying the properties of the CPU.
+    }
+}
+
+mod imp {
+    const NOTE: &str = "confirm hardware requirements at \
+        <https://docs.near.org/docs/develop/node/validator/hardware>.";
+
+    #[cfg(not(target_arch = "x86_64"))]
+    pub fn ensure_cpu_compatibility() {
+        eprintln!("error: processor running a x86_64 instruction set is required.\nnote: {}", NOTE);
+        std::process::exit(1);
+    }
+
+    #[cfg(target_arch = "x86_64")]
+    pub fn ensure_cpu_compatibility() {
+        // SAFETY: this is “just” a CPUID. Some very ancient chips don't support this instruction
```

Comment:
Not only chips, it's known that emulators (at least qemu and macOS Rosetta) doesn't provide cpuid compatible with at least Rust AVX checks.

Reply:
Do you have any more information? [Rosetta does not support AVX at all](https://developer.apple.com/documentation/apple-silicon/about-the-rosetta-translation-environment), but there doesn't seem to be any indications suggesting that `cpuid` returns incorrect information.